linked_list/linked_list.py

```diff
@@ -2,78 +2,140 @@
+        else:
+            new_node=Node(value)
+            new_node.next=self.head
+            self.head=new_node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        node=self.head
+        while node:
+            if self.head.value==value:
+                return True
+            node=self.head.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        counter=0
+        node=self.head
+        while node:
+            counter+=1
+            node=self.head.next
+        return counter
+
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        if self.length()<index:
+            return None
+        counter=0
+        node=self.head
+        while node:
+            if counter==index:
+                return self.head.value
+            node=self.head.next
+            counter+=1
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        if self.length()==0:
+            return None
+        counter=0
+        node=self.head
+        while node:
+            counter+=1
+            if self.head.next == None:
+                tail=Node(self.head)
+                return tail.value
+        return None
+        
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        if self.length()==0:
+            self.add_first()
+        else:
+            counter=0
+            node=self.head
+            while node:
+                counter+=1
+                if self.head.next == None:
+                    break
+            self.head.next=Node(value)
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        
+        node=self.head
+        max=self.head.value
+        while node:
+            if self.head.value>max:
+                max=self.head.value
+            node=node.next

```

Comment:
💫 This is great! This updates `node` to hold the next node in the linked list, and thus helps our loop make progress towards exiting! Can you apply this in some of your other methods to fix bugs with infinite loops you may be having??